GUI/Main/MainChangeset.cs

```diff
@@ -37,6 +37,8 @@ private void Changeset_OnConfirmChanges()
             // Using the changeset passed in can cause issues with versions.
             // An example is Mechjeb for FAR at 25/06/2015 with a 1.0.2 install.
             // TODO Work out why this is.
+            if (installWorker.IsBusy)
+                return;
```

Comment:
Hmm, I'm not completely sold on using `IsBusy` for this. Technically it still leaves us vulnerable to a race condition if we confirm that `IsBusy` is false and then another thread causes it to become true before we call `RunWorkerAsync`. What if we catch and ignore `InvalidOperationException`?

Reply:
Alternately, we could disable the button after the user clicks it.

Reply:
> Hmm, I'm not completely sold on using IsBusy for this. Technically it still leaves us vulnerable to a race condition if we confirm that IsBusy is false and then another thread causes it to become true before we call RunWorkerAsync.

What thread apart from the GUI thread could start it though, realistically?

> What if we catch and ignore InvalidOperationException?
Alternately, we could disable the button after the user clicks it.

In this case I prefer the catch solution. Disabling the button was my first attempt, but it didn't work out out at all, there are too many places where we hide and show and change the changeset tab. Sometimes the button needs to be re-enabled without the changeset changing (after failed installations), but not too early so it actually works, but also re-enable when the changeset changes...

Reply:
Another possibility, tell the button to handle double clicks separately:

- https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.controlstyles?view=netframework-4.7.2#System_Windows_Forms_ControlStyles_StandardDoubleClick
- https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.control.doubleclick?view=netframework-4.7.2
- https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.button.mousedoubleclick?view=netframework-4.7.2

![image](https://user-images.githubusercontent.com/1559108/109832940-78025980-7c06-11eb-9ac9-ba7f912cf62e.png)

Reply:
Wouldn't this still leave room for the exception to throw, if the user clicks slow enough that it isn't registered as double-click, but too fast for the GUI to switch to the new tab?